components/style/properties.mako.rs

```diff
@@ -5323,6 +5472,63 @@ impl ComputedValues {
         self.font.clone()
     }
 
+    // http://dev.w3.org/csswg/css-transforms/#grouping-property-values
+    pub fn get_used_transform_style(&self) -> computed_values::transform_style::T {
+        use computed_values::mix_blend_mode;
+        use computed_values::transform_style;
+
+        let effects = self.get_effects();
+
+        // TODO(gw): Add clip-path, isolation, mask-image, mask-border-source when supported.
+        if effects.opacity < 1.0 ||
+           !effects.filter.is_empty() ||
+           effects.clip.is_some() {
+           effects.mix_blend_mode != mix_blend_mode::T::normal ||
+            return transform_style::T::flat;
+        }
+
+        if effects.transform_style == transform_style::T::auto {
+            if effects.transform.is_some() {
+                return transform_style::T::flat;
+            }
+            if effects.perspective != computed::LengthOrNone::None {
+                return transform_style::T::flat;
+            }
+        }
+
+        // Return the computed value if not overridden by the above exceptions
+        effects.transform_style
+    }
+
+    pub fn get_transform_kind(&self) -> TransformKind {
+        // Check if the transform matrix is 2D or 3D
+        if let Some(ref transform_list) = self.get_effects().transform {
+            for transform in transform_list {
+                match transform {
```

Comment:
Not in a way that requires a 3d transform layer (unless perspective is present). But I'll update the comment / code to be more clear what it's checking for.